teps/0057-windows-support.md

```diff
@@ -0,0 +1,154 @@
+
+### Goals
+- Allow Tekton TaskRuns to selectively run on Windows nodes
+- Allow a TaskRun to use nodeSelector to request this feature
+- Core Tekton components (Operator, CLI and Dashboard) must be able to recognise and communicate with Windows nodes in the same way they do Linux nodes
+- Tasks on the Catalog and Hub should be able to indicate the requirement for a Windows node to run it
+- All current (Linux-only) workflows should remain unchanged
+
+### Non-Goals
+- The ability to run core Tekton components on Windows nodes is not a goal of this TEP
+    - It is expected that users will have a mix of Linux and Windows nodes and that the core Tekton controller components will run on Linux nodes
+- Tekton will not be responsible for determining whether a Pod for a TaskRun should be created on a Windows node. This will be an opt-in feature determined via a nodeSelector.
+
+
+<!-- ### Use Cases (optional) -->
+
+<!--
+Describe the concrete improvement specific groups of users will see if the
+Motivations in this doc result in a fix or feature.
+
+Consider both the user's role (are they a Task author? Catalog Task user?
+Cluster Admin? etc...) and experience (what workflows or actions are enhanced
+if this problem is solved?).
+-->
+
+## Requirements
+Tekton should be able to:
+- Recognise and communicate with Windows nodes in the cluster
+- Allow TaskRuns to specify the need to run on a Windows node
+- Create Pods for TaskRuns on Windows nodes
+- Provide full support for all Tekton features to Windows based TaskRuns
+- Manage Pipelines containing a mix of Windows and Linux based Tasks
+
+Workflows that do not require Windows support should not be changed in any way.
+
+Tekton Hub and Catalog should allow a Task to indicate that it requires a Windows node to run, if this is absent, the default behaviour will be to assume Linux so that no change to existing Linux workflows is required.
+
+## Proposal
+Extend Tekton so TaskRuns can request, and be scheduled, to run on Windows nodes. This will require (at least) reworking the following Tekton components:
+- Entrypoint binary
+- All internal support images used for resource management
+- File paths used for step ordering, either via polling or watching files
+- Script mode
+
+### Notes/Caveats (optional)
+In order to have Windows nodes run TaskRuns there are some elements of Tekton that will need reworking to be compatible with that OS. Some of these have been identified in the [issue](https://github.com/tektoncd/pipeline/issues/1826) but it is almost certain that other elements will need work in order to work. 
+
+### Risks and Mitigations
+- The burden of supporting the Windows platform may be quite high
+- Windows containers may introduce some security vulnerabilities
```

Comment:
Thanks for pointing these out. They do make me quite nervous 😄 

Can you think of any mitigations? We could explicitly state that while Tekton might _work_ on Windows, but that it isn't explicitly _supported_ for now, and until we get a bit more maturity we don't recommend it for real-world usage?

Reply:
Unfortunately the security issues around windows containers are not going to go away until a real fix for windows containers is implemented. Until then I think we should see use of windows containers as a 'here be dragons' situation - be careful using them, and absolutely don't run any untrusted workloads in them.

I'd agree that we should avoid recommending it for real-world usage until there is a decent degree of maturity present.

Reply:
are there specific vulnerabilities we're concerned about?